### ISSUE TITLE
Cannot 'export default' abstract, ambient class or interface

### ISSUE THREAD

**Issue by zhuravlikjb** [Microsoft/TypeScript#3792]
```
export default abstract class B {

}
```

By the way, the same with

```
export default declare class B {

}
```

Seems to be a parser bug

**Comment by DanielRosenwasser** [comment]
Under what situations would one need to make an ambient class a default export?

**Comment by mhegazy** [comment]
These two are by desing. We have felt that the `export default` syntax is already long enough, so no modifiers are needed. for a workaround use:
```ts
declare class  C {}
export default C;
```

**Comment by danquirk** [comment]
We should just allow these. It's a strange inconsistency, the workaround is _more_ verbose, and the error gives you no help getting to the workaround.

**Comment by DanielRosenwasser** [comment]
Yeah, discussed with @danquirk and @RyanCavanaugh offline; in a `.d.ts` file, a default exported ambient class makes sense.

I don't think it's that unreasonable to support the original syntax.

**Comment by mhegazy** [comment]
should also cover interfaces (https://github.com/Microsoft/TypeScript/issues/3914):
```ts
export default interface User {
    wpUserID: string;
}
```

**Comment by RyanCavanaugh** [comment]
And namespaces: `export default namespace foo { ... }` (#7407)

**Comment by RyanCavanaugh** [comment]
Accepting PRs for this.

**Comment by DanielRosenwasser** [comment]
Does this mean a namespace can be anonymous?

```ts
export default namespace {
    export var foo = 100;
}
```

What about an interface?

```ts
export default interface {
    foo: number;
}
```

**Comment by RyanCavanaugh** [comment]
No other changes except for allowing the `export default ... ... ` modifier combinations that were previously illegal.

**Comment by seanchas116** [comment]
I would prefer support for `export default enum` and `export default const enum` too. (#3320)

**Comment by clavecoder** [comment]
I'm OK with leaving it as is if the reasoning is that interfaces, et al, are not vaid **EcmaScript** and #3917 is implemented.

**Comment by mhegazy** [comment]
Unsupported `export default` patterns:

* [X] `export default abstract class C { }` Added by #14657
* [ ] `export default declare class C { }` 
* [X] `export default interface I { }` Added by #16040
* [ ] `export default enum E { }`
* [ ] `export default enum E { }`
* [ ] `export default const enum E { }`
* [ ] `export default namespace N { }`

**Comment by anstarovoyt** [comment]
@mhegazy what about `export default type foo = {}` ?

**Comment by simonbuchan** [comment]
@mhegazy for symmetry with JS export forms, consider (strawman syntax) `export default as ...`:

JS | TS
---|---
`export const x = 123` | `export type T = number`
`export { x }` | `export { T }`
`export class Foo {}` | `export interface Foo {}`
`export default class Foo {}` | `export default interface Foo {}`
`export default 123` | `export default as number`

**Comment by mhegazy** [comment]
not sure what that means.. 
```ts
export default as number
```
but if it is a .d.ts, we chose not to have a special syntax for this, and the recommendation here is to use 
```ts
declare cosnt _t: number;
export default _t;
```

**Comment by simonbuchan** [comment]
Yeah, sorry, it is a bit ambiguous. Currently, you can sort of export a type as default with:

```ts
type T = number;
export default T;
```

By analogy of `const` ~= `type`:

```ts
export const x = 123;
export type T = number;
export default x;
export default T;
export default 123;
export default ???; // can't put type expressions here, only type names, due to ambiguity
```

The motivating case would would be something like:

```ts
// Action.ts
export default as
  | { type: 'login', name: string, password: string }
  | { type: 'logout' }
  | ...
```

where currently you would need to bury the export at the bottom of the file.

I picked `export default as <type>` as strawman simply because `as` already suggests a type on RHS, but it doesn't really imply the right thing, you're right. Perhaps `export type default <type>`?

**Comment by alvis** [comment]
Really anticipate #18628! Thanks @NaridaL. 👏👏 👏
